### ISSUE TITLE
Lint on `.as_ptr().cast_mut()`

### ISSUE THREAD

**Issue by stegaBOB** [rust-lang/rust-clippy#14651]
### What it does

Creating a `*mut T`  from a `&T` is almost always bad. This lint would warn against doing so and potentially suggest calliing `.as_mut_ptr()` if applicable. There are many types in std that have a `.as_ptr ()` method. This lint would apply to all of them if possible.

### Advantage

- Prevent possible UB from dereferencing a `*mut` ptr from a shared reference into a mutable reference
- Warn the user if they are doing `.as_ptr().cast_mut()` on a shared reference since de-referencing the resulting pointer would be UB

### Example

```rust
let mut data = vec![1u8; 10];
let ptr: *mut u8 = some_data.as_ptr().cast_mut();
```

Could be written as:

```rust
let mut data = vec![1u8; 10];
let ptr: *mut u8 = some_data.as_mut_ptr();
```
